### ISSUE TITLE
"..Property '[Symbol.iterator]' is missing..." error started to appear after latest release

### ISSUE THREAD

**Issue by ZuBB** [TypeStrong/atom-typescript#1329]
Hi there!

After we got latest release with fix of atom's 1.19 changes I started to got lots of errors of where subject phrase is included

```
Argument of type 'typeof AtvChangeSchemaModalController' is not assignable to parameter of type 'Injectable<IControllerConstructor>'. Type 'typeof AtvChangeSchemaModalController' is not assignable to type '(string | (new (...args: any[]) => IController) | ((...args: any[]) => void | IController))[]'. Property '[Symbol.iterator]' is missing in type 'typeof AtvChangeSchemaModalController'.
```

I am not getting this error neither in vscode nor in terminal with webpack+tslint.

@guncha, @lierdakil do you have any idea(s) what can be wrong?

**Comment by lierdakil** [comment]
No, not really. First idea that comes to mind is you're missing `es6` in `lib` in `tsconfig.json` somewhere. If not, then if you could slap together a minimal example demonstrating the bug, we could investigate further.

**Comment by ZuBB** [comment]
hi @lierdakil! 

yea, we do not have an `es6` keyword in tsconfig. I am afraid I cant just change that. that kind of changes are being driven by another person and should be same across all projects in our company.

but I still would like to know how it works for other editors..

**Comment by lierdakil** [comment]
Shooting in the dark here, but maybe different TypeScript versions are used by atom-typescript and everything else? Check if you have different project-local and global installations maybe? (`node_modules/.bin/tsc --version` vs `tsc --version`)

**Comment by lierdakil** [comment]
Well, I'm out of ideas for now.

Re-reading the error message again, it reads as if
```typescript
Injectable<IControllerConstructor> = (string | (new (...args: any[]) => IController) | ((...args: any[]) => void | IController))[]
```
notice that `Injectable` is an array,
and it seems you're trying to assign `AtvChangeSchemaModalController` to that, which isn't an array.

It all looks pretty mysterious, especially if I don't see any code. Again, if you could slap together something that shows similar behavior, that would make things somewhat easier.

**Comment by ZuBB** [comment]
I will try to get to this a bit closer one more time

**Comment by ZuBB** [comment]
I have searched a bit and found next thing https://github.com/Microsoft/TypeScript/issues/12707#issuecomment-265518894

since I do not have this issue with webpack/vscode but have with atom is that possible that your package somehow sets target to es6?

ps: here is my tsconfig
```
...
  "compileOnSave": false,
  "compilerOptions": {
    "lib": ["dom", "es5", "es2015.promise", "es2015.collection", "es2015.iterable"],
    "target": "es5",
    "stripInternal": false,
    "sourceMap": true,
    "module": "commonjs", // "es6" value conflicts with ng-annotate :((
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "noEmitHelpers": true,
    "importHelpers": true,
    "typeRoots" : ["./unity-core/node_modules/@types"],
    "baseUrl": ".",
...
```

**Comment by lierdakil** [comment]
atom-typescript doesn't touch tsconfig. All it does, basically, is spawns a ts language server, feeds it data and displays its responses. So not likely.

I should point out that `Symbol.iterator` is not defined in ES5. Since, again, I don't see code, I can't really make informed statements about the code. https://github.com/Microsoft/TypeScript/issues/4031 might be relevant. I see you have `es2015.iterable` in `lib`. Which either means you're using a shim (which TS might or might not know about), or something else is going on behind the scenes.

**Comment by ZuBB** [comment]
ok, I have made some test and I found that it for sure started to happen with 11.0.6 -> 11.0.7 update.

here is what I did
* quit my current atom instance
* rename ~/.atom to .atom-b
* download atom 1.18
* unzip it to tmp location and start
* install typescript (only 11.0.6 are allowed because of https://github.com/TypeStrong/atom-typescript/commit/333f4049149c2f876bbde377a6a1ee39d52662f9#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L17) and linter packages
* reload atom window to make installed plugins work
* open files which have strange errors (one that is stated as subject, and there are anothers)

with this setup atom@1.18 + typescript@11.0.6 all is OK

then I did next
* manually installed typescript@11.0.7 via `apm install atom-typescript@11.0.7`
* reload atom window to make installed plugins work

with this setup I started to get those strange errors

**Comment by ZuBB** [comment]
here are screens with errors in case it may help

<img width="1152" alt="screen shot 2017-09-14 at 14 11 00" src="https://user-images.githubusercontent.com/597016/30426962-ae43ff58-9956-11e7-9450-6b7ba82f95f9.png">
<img width="1140" alt="screen shot 2017-09-14 at 14 10 15" src="https://user-images.githubusercontent.com/597016/30426961-ae3ffc28-9956-11e7-9d2d-f11b000cb0bd.png">

**Comment by ZuBB** [comment]
ok, I believe I know what causes this bug. its this line https://github.com/TypeStrong/atom-typescript/commit/333f4049149c2f876bbde377a6a1ee39d52662f9#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L70

we used to use typescript@2.3.4 and webpack did not produce any errors. but when I manually updated to to typescript@2.4.2 as you did I started to get those errors.

I think real fix would be make atom-typescript use typescript that is defined in projects's `package.json` by default and if not found -- fallback to own instance.

@lierdakil @guncha, what do you think?

**Comment by lierdakil** [comment]
It *should already* use package-specific instance of typescript =\ If it can find one, that is.

More specifically, it tries to resolve `typescript/bin/tsserver` Node module for currently open file using [node resolve algorithm](https://nodejs.org/api/modules.html#modules_all_together). You need to have typescript module installed of course.

**Comment by ZuBB** [comment]
ah, then it makes things even more clear :(

our main package.json is in another dir and all was OK till your version and our matched. with 11.0.7 it stopped to be equal.

is there any way that I can specify custom path for `node_modules` folder where it looks for `typescript` package?

**Comment by lierdakil** [comment]
Symlink `node_modules`? Set `NODE_PATH` environment variable? Note that node resolve algorithm doesn't care about `package.json`, it only looks for `node_modules` subdirectory.

Also note that node resolve algorithm searches for modules not only in `$cwd/node_modules`, but also in `$cwd/../node_modules`, `$cwd/../../node_modules`, etc.

**Comment by ZuBB** [comment]
ok, I got your idea. 

unfortunately it does not help

**Comment by ZuBB** [comment]
@lierdakil @guncha, just for your information: I had some discussion on it with my colleagues who use different IDES and it turned out that Webstorm has an option to set custom path for dir with typescript package

**Comment by lierdakil** [comment]
Okay, I did find a small bug in module resolution code, which kinda breaks with multiple directories in NODE_PATH. That is fixed by v11.0.9.

FWIW, it is feasible to "set custom path", that atom-typescript will fallback to, if that's what you want to do. It might be a bit arcane though.

1. Add this to your init script (Edit → Init Script...):

    On *nix:
    ```coffee
    process.env.NODE_PATH += ":/path/to/node_modules" # that contains typescript module
    ```
    (note string starts with colon)

    On Windows:
    ```coffee
    process.env.NODE_PATH += ";C:\\path\\to\\node_modules" # that contains typescript module
    ```
    (note string starts with semicolon)

2. Restart Atom (or `window:reload`)

3. Profit?

Personally, I would just `npm i typescript@2.3` near `tsconfig.json`, if that makes sense. Not entirely sure about your project layout, so can't judge if it's feasible.

**Comment by ZuBB** [comment]
hey @lierdakil thanks for your response. yes, I already did what you suggest as 2nd option
